Fixes BUG: `outer_join` generates LEFT join instead of full outer

scottcode opened the pull request:
Fixes issue #1772

When trying to create a full outer join using the `outer_join` method on a `TableExpr`, the resulting SQL is actually a `LEFT OUTER JOIN`.

```
iconn = ibis.sql.postgres.client.PostgreSQLClient(...)

iconn.con.execute("DROP TABLE IF EXISTS test_left;")
iconn.con.execute("CREATE TABLE test_left(id_left integer, value_left integer);")
iconn.con.execute("DROP TABLE IF EXISTS test_right;")
iconn.con.execute("CREATE TABLE test_right(id_right integer, value_right integer);")

left = iconn.table('test_left')
right = iconn.table('test_right')

joined = left.outer_join(right, left.id_left == right.id_right)

print(joined.compile())
# SELECT t0.id_left, t0.value_left, t1.id_right, t1.value_right
# FROM test_left AS t0 LEFT OUTER JOIN test_right AS t1 ON t0.id_left = t1.id_right

'full outer' in str(joined.compile()).lower()
# False
```

The root cause of it is that the method in `ibis.sql.alchemy._AlchemyTableSet.get_result()` uses `sqlalchemy.sql.selectable.FromClause.outerjoin()` with just `result.outerjoin(table, onclause)`. However, starting in sqlalchemy version `1.1.0` (circa 2016) they added a keyword argument `full` with default `False`. To fix this we need to change the following:

Current:
```
elif jtype is ops.OuterJoin:
    result = result.outerjoin(table, onclause)
```

Proposed fix:
```
elif jtype is ops.OuterJoin:
    result = result.outerjoin(table, onclause, full=True)

```

cpcloud commented:
Thanks for the PR and for hunting down the issue! I will try to review this over the weekend.

cpcloud commented on ibis/tests/all/test_outer_join.py:
```diff
@@ -0,0 +1,41 @@
+import os
```
Thanks for writing a unit test!

The way that we typically write unit tests is using a slightly different style, namely that of [`pytest`](https://docs.pytest.org/en/latest/). `pytest` uses a more functional style and avoids mixing up fixtures and test cases. Another nice thing is that it has a lot of builtin fixtures for common tasks like generating a temporary file.

Do you want to try rewriting this test in that style?

Here's a skeleton to get you started:

```python
@pytest.fixture
def client(tmpfile):  # tmpfile is passed in by pytest when the test runs
    con = ibis.sqlite.connect(str(tmpfile), create=True)
    # setup tables ...
    return con


@pytest.fixture
def left(client):
    return client.table('left_table')


@pytest.fixture
def right(client):
    return client.table('right_table')


def test_outer_join(left, right):
    ...
```

Another thing is that you should put this test in `ibis/sql/sqlite/tests/test_functions.py`.

After addressing those two things, this PR will be good to go!

scottcode replied:
Sounds great! I'm new to `pytest`, so thanks for the skeleton. I'll rework it and get back to you.

scottcode replied:
@cpcloud, I re-did the my test in the pytest fashion. Instead of doing so in SQLite, however, I did it in the higher-level `test_sqlalchemy.py`, because the issue wasn't specific to SQLite but to anything based on `sqlalchemy`. Along the way I discovered some issues with the MockConnection and the `test_joins` method, so I fixed and refactored those. 

These changes are in commit:  
scottcode:fix_full_outer_join  29b4a5106a3043bc

I noticed some changes in the upstream master since my first version of pull request. Do I need to make a new rebased pull request branch, or can you rebase when accepting the request? I haven't encountered this situation before in GitHub, so not sure what the convention is.

cpcloud commented:
@scottcode Rebasing isn't strictly necessary if your patch merges cleanly (i.e., without merge conflicts), but is definitely necessary if a PR has been up for a while to make sure it doesn't break master.

In this case you should probably rebase since there's been some testing changes and you want to make sure that you pick those up in testing (I don't think they're related, but better to be safe here).

I can also merge master back into your PR branch which will kick off another CI run.

Related to CI you have some lint failures that will prevent further testing: https://circleci.com/gh/ibis-project/ibis/7437?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

You can run those on your development machine by typing `make lint` on the command line to avoid having to wait for CI to spin up just to get linting errors.

Another thing that's helpful is if you install `pre-commit` and use that. It will run all the upstream static checks that we run in CI to make sure style and linting conform every time you make a commit. If your code doesn't pass muster then the commit will not succeed.

To use pre-commit run the following:

```sh
pip install pre-commit
cd wherever/you/cloned/ibis
pre-commit install
```

scottcode commented:
@cpcloud Thanks for all the pointers. I fixed the linter errors and rebased on master, creating a new branch and pull request (#1788). I wasn't sure if there was a way to incorporate that new branch into this same PR, so that's why I did it that way. Thanks!

cpcloud commented:
For future reference you can run `git pull --rebase upstream/master` (as long as you have a remote named `upstream` referring to `ibis-project/ibis`) to pull in the latest changes from upstream. You can then run `git push --force` to force push to your remote.
